app/utils/use-fetch-data.ts

```diff
@@ -1,41 +1,125 @@
-import React from "react";
+import { useCallback, useEffect, useRef, useState } from "react";
+import { stringifyVariables } from "urql";
 
 type Status = "idle" | "fetching" | "success" | "error";
 
-type UseFetchDataOptions = {
+export type UseFetchDataOptions<TDefault = undefined> = {
   enable?: boolean;
   initialStatus?: Status;
+  defaultData: TDefault;
 };
 
-export const useFetchData = <T>(
-  queryFn: () => Promise<T>,
-  options: UseFetchDataOptions = {}
+type QueryKey = string[];
+
+type QueryCacheValue<T> = {
+  data: T | null;
+  error: Error | null;
+  status: Status;
+};
+
+class QueryCache {
+  cache: Map<string, QueryCacheValue<unknown>>;
+  listeners: [string, () => void][];
```

Comment:
Same idea but here the QueryCache only notifies subscribers for the key, while Observable fire for every listeners. Since there is little code, I am not sure it is needed to share the observable behavior here.